consensus/core/src/authority_node.rs

```diff
@@ -661,4 +511,200 @@ mod tests {
+
+        const NUM_TRANSACTIONS: u8 = 15;
+        let mut submitted_transactions = BTreeSet::<Vec<u8>>::new();
+        for i in 0..NUM_TRANSACTIONS {
+            let txn = vec![i; 16];
+            submitted_transactions.insert(txn.clone());
+            authorities[i as usize % authorities.len()]
+                .transaction_client()
+                .submit(vec![txn])
+                .await
+                .unwrap();
+        }
+
+        for receiver in &mut output_receivers {
+            let mut expected_transactions = submitted_transactions.clone();
+            loop {
+                let committed_subdag =
+                    tokio::time::timeout(Duration::from_secs(1), receiver.recv())
+                        .await
+                        .unwrap()
+                        .unwrap();
+                for b in committed_subdag.blocks {
+                    for txn in b.transactions().iter().map(|t| t.data().to_vec()) {
+                        assert!(
+                            expected_transactions.remove(&txn),
+                            "Transaction not submitted or already seen: {:?}",
+                            txn
+                        );
+                    }
+                }
+                assert_eq!(committed_subdag.reputation_scores_desc, vec![]);
+                if expected_transactions.is_empty() {
+                    break;
+                }
+            }
+        }
+
+        // Stop authority 1.
+        let index = committee.to_authority_index(1).unwrap();
+        authorities.remove(index.value()).stop().await;
+        sleep(Duration::from_secs(5)).await;
+
+        // now create a new directory to simulate amnesia. The node will start having participated previously
+        // to consensus but now will attempt to synchronize the last own block and recover from there.
+        let (authority, receiver) = make_authority(
+            index,
+            TempDir::new().unwrap(),
+            committee.clone(),
+            keypairs,
+            network_type,
+        )
+        .await;
+        output_receivers[index] = receiver;
+        authorities.insert(index.value(), authority);
+        sleep(Duration::from_secs(5)).await;
+
+        // Stop all authorities and exit.
+        for authority in authorities {
+            authority.stop().await;
+        }
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_amnesia_failure(
+        #[values(ConsensusNetwork::Anemo, ConsensusNetwork::Tonic)] network_type: ConsensusNetwork,
+    ) {
+        telemetry_subscribers::init_for_testing();
+
+        let occurred_panic = Arc::new(Mutex::new(None));
+        let occurred_panic_cloned = occurred_panic.clone();
+
+        let default_panic_handler = std::panic::take_hook();
+        std::panic::set_hook(Box::new(move |panic| {
+            let mut l = occurred_panic_cloned.lock().unwrap();
+            *l = Some(panic.to_string());
+            default_panic_handler(panic);
+        }));
```

Comment:
If anyone has a better idea how to catch the underlying panics please let me know. Unfortunately the panic from the underlying thread is not propagating automatically up to the root state of the test and it rather makes it pass, although there is the panic.